hummingbot/market/kucoin/kucoin_api_order_book_data_source.py

```diff
@@ -0,0 +1,300 @@
+    Dict,
+    List,
+    Optional
+)
+import re
+import time
+import ujson
+import websockets
+from websockets.exceptions import ConnectionClosed
+
+from hummingbot.core.utils import async_ttl_cache
+from hummingbot.core.utils.async_utils import safe_gather
+from hummingbot.core.data_type.order_book_tracker_data_source import OrderBookTrackerDataSource
+from hummingbot.core.data_type.order_book_tracker_entry import OrderBookTrackerEntry
+from hummingbot.core.data_type.order_book_message import OrderBookMessage
+from hummingbot.core.data_type.order_book import OrderBook
+from hummingbot.logger import HummingbotLogger
+from hummingbot.market.kucoin.kucoin_order_book import KucoinOrderBook
+
+TRADING_PAIR_FILTER = re.compile(r"(BTC|ETH|USDT)$")
+
+SNAPSHOT_REST_URL = "https://api.kucoin.com/api/v1/market/orderbook/level2_100"
+DIFF_STREAM_URL = ""
+TICKER_PRICE_CHANGE_URL = "https://api.kucoin.com/api/v1/market/allTickers"
+EXCHANGE_INFO_URL = "https://api.kucoin.com/api/v1/symbols"
+
+
+class KucoinAPIOrderBookDataSource(OrderBookTrackerDataSource):
+
+    MESSAGE_TIMEOUT = 30.0
+    PING_TIMEOUT = 10.0
+
+    _kaobds_logger: Optional[HummingbotLogger] = None
+
+    @classmethod
+    def logger(cls) -> HummingbotLogger:
+        if cls._kaobds_logger is None:
+            cls._kaobds_logger = logging.getLogger(__name__)
+        return cls._kaobds_logger
+
+    def __init__(self, symbols: Optional[List[str]] = None):
+        super().__init__()
+        self._symbols: Optional[List[str]] = symbols
+        self._order_book_create_function = lambda: OrderBook()
+
+    @classmethod
+    @async_ttl_cache(ttl=60 * 30, maxsize=1)
+    async def get_active_exchange_markets(cls) -> pd.DataFrame:
+        """
+        Returned data frame should have symbol as index and include usd volume, baseAsset and quoteAsset
+        """
+        async with aiohttp.ClientSession() as client:
+
+            market_response, exchange_response = await safe_gather(
+                client.get(TICKER_PRICE_CHANGE_URL),
+                client.get(EXCHANGE_INFO_URL)
+            )
+            market_response: aiohttp.ClientResponse = market_response
+            exchange_response: aiohttp.ClientResponse = exchange_response
+
+            if market_response.status != 200:
+                raise IOError(f"Error fetching Kucoin markets information. "
+                              f"HTTP status is {market_response.status}.")
+            if exchange_response.status != 200:
+                raise IOError(f"Error fetching Kucoin exchange information. "
+                              f"HTTP status is {exchange_response.status}.")
+
+            market_data = await market_response.json()
+            exchange_data = await exchange_response.json()
+            
+            attr_name_map = {"baseCurrency": "baseAsset", "quoteCurrency": "quoteAsset"}
+
+            trading_pairs: Dict[str, Any] = {item["symbol"]: {attr_name_map[k]: item[k] for k in ["baseAsset", "quoteAsset"]}
```

Comment:
I am getting a `KeyError` here.

I think you meant to write
> trading_pairs: Dict[str, Any] = {... for k in ["**baseCurrency**", "**quoteCurrency**"]}

Reply:
Corrected, thanks.